### ISSUE TITLE
NEED to rework the Mods submission / maintenance workflow

### ISSUE THREAD

**Issue by NOCanoa** [zen-browser/theme-store#928]
Ye we need a better workflow to manage all the mods

**Comment by RobotoSkunk** [comment]
I think the idea of using a monorepo to host everything is good to keep everything open source, but a proper implementation should be using a hosted server where people upload their mods individually without waiting for a mass mods approbation

**Comment by rsiebertdev** [comment]
I think it'd be better to archive the monorepo. Instead, let's take a *similar* approach to the Firefox or Chrome Extensions platforms. Zen already has a modding gallery, but instead of submitting mods as issues in this monorepo, developers should be able to log in to the modding platform with their GitHub accounts and upload mods directly.

Rather than uploading *files* to the platform, you'd simply provide a link to their mod’s GitHub repository. This ensures all mods remain open-source and stay properly credited.

To update a mod, you'd navigate to your Mod Details and **create a new version** by entering details like the version number and description to be displayed on the change logs. Once submitted, the backend queues the update, pulling the latest commit from your repository. It handles the rest, minifying, packaging and running checks where necessary, finally publishing the update to the database viewable in the gallary or pulled by browsers (like it is at the moment).

¯\\\_(ツ)_/¯

**Comment by WillsterJohnson** [comment]
mods as issues is kind of antithetical to the idea of a git repo imo. it should either be via PR like some software user repositories, or via a web platform like browser addons.

Given that it's just CSS and there's nothing executable, I think that a user repository design is best. There's no need to do the kind of safeguarding done in browser extensions (CSS isn't script), and it's completely possible to have the customer-facing web interface merely reference this repo rather than building and entire system for doing exactly what a public git repo does but with objectively worse version control.

Having a git repo means mod authors can publish incremental updates to their mods just like any other software, those changes can be tracked, and any build processes can happen in a known and standardised environment that's guaranteed to be the same on everyone's machine. The enterprising Zen user can even try out those new versions before they're merged in.

For example, it would be (almost) free to support sass/less/stylus in exactly the same way for every mod author, whereas a web upload interface would require that author to set up their own configs as well as a github workflow to run build & deploy (which would bypass the web interface anyways). nb; I'm not saying do this, I'm plucking an example out of thin air to demonstrate the affordance of a git repo over a web interface.

As for downloads, favorites, ratings, comments, etc etc, just do that exactly the same as in the web upload model, but tell the server that the UUID points to a path in the HEAD of this repo (or cached release files) rather than someplace in it's database.

A great example of just how much *\~magic\~* you can pump into a user repository while still having a customer-facing directory of stuff is [nixpkgs](https://github.com/NixOS/nixpkgs).

I also think that using directory names as the UUID isn't great. See nixpkgs above, the 'id' (in their case, a human-readable name) is defined by a particular field in the configuration file. I think it would be best to have a directory structure of either `/themes/<mod name>/` or `/themes/<github username>/<mod name>/`, and require a `manifest.json` containing the basic info including the UUID. Heck, with the directory structures above you can completely forego a UUID at all; files in the same directory are guaranteed to be unique amongst themselves.

Ultimately, whatever the solution is, it is not using issues as a means of code contribution.

---

Side note;

If I submit a PR to this repo now, will it be rejected on principle or will it be given a chance?

I mean no disrespect and hope you can understand when I say; I completely refuse to interact with this issue-as-a-code-contribution nonsense.

**Comment by lindongbin** [comment]
It is recommended to submit CSS freely like Greasy Fork

**Comment by CosmoCreeper** [comment]
Another idea would be to go to the [zen-browser.app/mods](https://zen-browser.app/mods) site and be able to have an account (you would probably just sign in with your GitHub account) which you can use to update and publish your mods. 

**Publish a mod:** Upon attempting to publish a mod, the website could let you know immediately if your image is the inappropriate size or if certain details are incorrect. If all properties are valid and everything is set to go, you could hit a publish button which would automatically create a pull request in the current format that is used.

**Update a mod:** Upon attempting to update a mod, it would basically do the same thing as publishing except fields are automatically filled out unless you replace it with something else and you would need a theme.json file.

**Automatic repository folder and root detection:** One *amazing* feature that should definitely be implemented is the ability to just submit the link to your repository which contains your mod or the folder which contains your mod, and, based on the format everything should be in, it would automatically find all the files and get straight to publishing or updating your mod. This feature could also be checked on or off if you don't have a GitHub repository and just want to jump straight to publishing your mod.

**Comment by WillsterJohnson** [comment]
> Another idea would be to go to the [zen-browser.app/mods](https://zen-browser.app/mods) site and be able to have an account (you would probably just sign in with your GitHub account) which you can use to update and publish your mods.

I think doing that would be good if you can link a mod to a GitHub repo. This could even forego storing the mod in a database or in this git repo;

**Creation**: a 'link github repo' button takes you through the GitHub auth flow for your chosen repository.

**Updates**: Just create a release and push a `mods` folder to the artifacts. The content of that folder is your mod. The Zen Mods server will periodically (daily, weekly) poll repos for new releases. If one is detected, that mod has an update available.

**Validation**: The Zen Mods server can refuse to broadcast a mod if the `mods` release artifact contains anything other than what is expected. The developer will be notified in their Zen Mods dashboard.

**Installation**: Zen itself will be given the GitHub repo after validation, and can go directly to GitHub to download the mod. Honestly it may make more sense to do the validation step in Zen itself, as the release files could have been modified since the server validated them.

Whatever the solution, the important thing is to not create a dashboard that's as awful as the Mozilla or Chrome extension dashboards. Mod developers are writing code, let them use code tools and ecosystems.

**Comment by CosmoCreeper** [comment]
> I think doing that would be good if you can link a mod to a GitHub repo. This could even forego storing the mod in a database or in this git repo;

Good ideas! The other thing I noticed is that moderators of this repository tend to update mods a lot faster than publish mods. This automation process could make everything a lot faster regardless of publishing or updating.

**Comment by WillsterJohnson** [comment]
> > I think doing that would be good if you can link a mod to a GitHub repo. This could even forego storing the mod in a database or in this git repo;
> 
> Good ideas! The other thing I noticed is that moderators of this repository tend to update mods a lot faster than publish mods. This automation process could make everything a lot faster regardless of publishing or updating.

Agreed. It could also take a bit of work off their plate (once its all implemented of course), which I'm sure they would appreciate.

Perhaps this new Zen mods store app could be open source so folks can contribute to it. Eg, I would be interested in writing the GitHub integration when the app is ready for it if I've got time <sup>and if the app is in a language I know</sup>.

Though I think that before we get too far ahead of ourselves we should let the Zen team weigh in. They've been quiet on this issue so far, probably busy with the many new browser releases they've been making of late (love that btw!). Maybe a few fresh ideas will coax them in lol

**Comment by CosmoCreeper** [comment]
> I would be interested in writing the GitHub integration when the app is ready for it if I've got time and if the app is in a language I know.

Yes I'm working on my own unofficial mods store right now. I doubt it would be used by many people, but it would definitely help me out in the long run. The thing I'm working on will have the store built-in to the zen mods spot and will utilize fx-autoconfig to do a lot of the stuff. It will have auto-updating and people that add stuff to the repository will only ever have to create a pull request for adding, not updating. I am also going to add the ability to add unpublished mods in by just feeding the program a URL that it can update from.

Thinking of calling it "Sine".

**Comment by greeeen-dev** [comment]
Doing the same here for [Silkthemes](https://github.com/Silkbrush/theme-store), we've already got some mods submitted and we're building a frontend for it now

**Comment by CosmoCreeper** [comment]
> Doing the same here for [Silkthemes](https://github.com/Silkbrush/theme-store), we've already got some mods submitted and we're building a frontend for it now

Yeah the Zen Explorer. Yours is probably gonna be more permanent. Mine is probably just gonna be another project for me.

**Comment by CosmoCreeper** [comment]
@aaronedev regarding the side projects?

1. Zen Mods is still alive and well, its just that new mods aren't really getting accepted cause the dev is busy.
2. My project, Sine, succeeded and is an alternative/improvement to Zen Mods.
3. Zen Explorer or Silkthemes is dead now that Sine has fulfilled its purpose of being compatible with all Firefox-based browsers on v2.

**Comment by aaronedev** [comment]
@CosmoCreeper ah ok i see i though maybe new issues were on halt due to this discussion maybe. Also i wondered if something is wrong with my issue. However, if you say the dev manually needs to accept it, then i think i just gotta wait 😅

**Comment by mr-cheffy** [comment]
What’s the difference between the current mods and Sine? I haven’t really used it myself, to be honest. The only thing that will never be merged is support for JS mods, mainly for security and privacy reasons. I believe that a privileged process like the main window shouldn’t be injected with untrusted sources, it’s not something that should be tampered with.

**Comment by tretuttle** [comment]
@mr-cheffy can you provide an update to this in any way or refer me to your most up to date statement on this? 

I understand themes/mods are not your priority but I am offering to fill in the gaps where I can. I have a lot of hardware with Zen built on everything.

**Comment by mr-cheffy** [comment]
Mods currently are under a soft freeze until we do some sort of re-factor so they dont break so much between many consecutive updates
